quiche/src/stream.rs

```diff
@@ -421,11 +432,42 @@ impl StreamMap {
     /// to when an existing stream becomes writable (or stops being writable).
     ///
     /// If the stream was already in the list, this does nothing.
-    pub fn mark_writable(&mut self, stream_id: u64, writable: bool) {
+    pub fn mark_writable(
+        &mut self, stream_id: u64, priority_key: &Arc<StreamPriorityKey>,
+        writable: bool,
+    ) {
         if writable {
             self.writable.insert(stream_id);
+
+            self.writable_prioritized.insert(Arc::clone(priority_key));
         } else {
             self.writable.remove(&stream_id);
+
+            self.writable_prioritized_remove(priority_key);
+        }
+    }
+
+    fn writable_prioritized_remove(
+        &mut self, priority_key: &Arc<StreamPriorityKey>,
+    ) {
+        let mut c = {
+            let ptr = Arc::as_ptr(priority_key);
+            unsafe { self.writable_prioritized.cursor_mut_from_ptr(ptr) }
+        };
+
+        c.remove();
+    }
+
+    pub fn writable_prioritized_update(
+        &mut self, old: &Arc<StreamPriorityKey>, new: &Arc<StreamPriorityKey>,
+        writable: bool,
+    ) {
+        if old.writable.is_linked() {
+            self.writable_prioritized_remove(old);
+        }
+
+        if writable {
```

Comment:
If the old key was linked then we can assume the new one should be inserted too? This way we wouldn't need the additional `writable` argument.

Reply:
@LPardue it seems this comment still applies. Also, probably need to add the same logic for `readable` too.